File: drills/elixir/solutions/nc41_maximum_subarray.exs
defmodule Solution do
  def max_sub_array([]), do: 0

  # Kadane: at each position the best subarray ending here either extends the
  # one ending just before it or starts fresh. A running total that has gone
  # negative can only hurt whatever follows, so it is dropped.
  def max_sub_array([first | rest]) do
    {_here, best} =
      Enum.reduce(rest, {first, first}, fn n, {here, best} ->
        here = max(n, here + n)
        {here, max(best, here)}
      end)

    best
  end
end

File: drills/go/solutions/nc41_maximum_subarray.go
package main

func maxSubArray(nums []int) int {
	best, current := nums[0], nums[0]
	for _, n := range nums[1:] {
		// Extend the run, unless it has gone negative: then start over here.
		current = max(n, current+n)
		best = max(best, current)
	}
	return best
}
